test/src/main/java/org/springframework/security/test/support/JwtAuthenticationBuilder.java

```diff
@@ -0,0 +1,227 @@
+ */
+public class JwtAuthenticationBuilder<T extends JwtAuthenticationBuilder<T>> {
+
+	private static final String DEFAULT_NAME = "user";
+
+	private static final String[] DEFAULT_AUTHORITIES =  { "ROLE_USER" };
+
+	private static final String DEFAULT_TOKEN_VALUE = "test.jwt.value";
+
+	private static final String DEFAULT_HEADER_NAME = "test-header";
+
+	private static final String DEFAULT_HEADER_VALUE = "test-header-value";
+
+	private static final Map<String, Object> DEFAULT_HEADERS =
+			Collections.singletonMap(DEFAULT_HEADER_NAME, DEFAULT_HEADER_VALUE);
+
+	private static final String ROLE_PREFIX = "ROLE_";
+
+	protected String name;
+
+	protected final Set<String> authorities;
+
+	private boolean isAuthoritiesSet = false;
+
+	protected final Map<String, Object> claims = new HashMap<>();
+
+	protected Converter<Jwt, Collection<GrantedAuthority>> authoritiesConverter;
+
+	private String tokenValue = DEFAULT_TOKEN_VALUE;
+
+	private final Map<String, Object> headers = new HashMap<>();
+
+	public JwtAuthenticationBuilder() {
+		this.authoritiesConverter = new JwtGrantedAuthoritiesConverter();
+		name(DEFAULT_NAME);
+		this.authorities = new HashSet<>(Arrays.asList(DEFAULT_AUTHORITIES));
+	}
+
+	/**
+	 * @param jwt fully configured JWT
+	 * @return pre-configured builder
+	 */
+	public T jwt(final Jwt token) {
+		final Map<String, Object> claims = new HashMap<>(token.getClaims());
+		putIfNotEmpty(JwtClaimNames.IAT, token.getIssuedAt(), claims);
+		putIfNotEmpty(JwtClaimNames.EXP, token.getExpiresAt(), claims);
+
+		return claims(claims).tokenValue(token.getTokenValue()).name(token.getSubject()).headers(token.getHeaders());
+	}
+
+	public T tokenValue(final String tokenValue) {
+		this.tokenValue = tokenValue;
+		return downCast();
+	}
+
+	public T name(final String name) {
+		this.name = name;
+		return downCast();
+	}
+
+	public T authority(final String authority) {
+		Assert.hasText(authority, "authority must be non empty");
+		if (!this.isAuthoritiesSet) {
+			this.authorities.clear();
+			this.isAuthoritiesSet = true;
+		}
+		this.authorities.add(authority);
+		return downCast();
+	}
+
+	public T authorities(final Stream<String> authorities) {
+		this.authorities.clear();
+		authorities.forEach(this::authority);
+		return downCast();
+	}
+
+	public T authorities(final String... authorities) {
+		return authorities(Stream.of(authorities));
+	}
+
+	public T authorities(final Collection<String> authorities) {
+		return authorities(authorities.stream());
+	}
+
+	public T role(final String role) {
+		Assert.hasText(role, "authority must be non empty");
+		Assert.isTrue(
+				!role.startsWith(ROLE_PREFIX),
+				"role must not be prefixed with " + ROLE_PREFIX + " (it is auto-added)");
+		return authority(ROLE_PREFIX + role);
+	}
+
+	public T roles(final Stream<String> roles) {
+		this.authorities.removeIf(a -> a.startsWith(ROLE_PREFIX));
+		roles.forEach(this::role);
+		return downCast();
+	}
+
+	public T roles(final String... roles) {
```

Comment:
Removed the existing roles cleanup, it now just adds to already registered ones as `UserRequestPostProcessor#roles` does.

Side note, I prefer the previous behavior: in case you have a method defining a custom authentication and want just a subset of the roles set for a certain test, then you are screwed.

Reply:
I don't think I quite followed. Why couldn't you do:

```java
MyCustomAuthentication authentication = new MyCustomAuthentication();
List<String> roles = authentication.getAuthorities().stream()
    .filter(this::predicate).map(Objects::toString).collect(toList());
JwtAuthenticationBuilder jwtAuthenticationBuilder = new JwtAuthenticationBuilder();
jwtAuthenticationBuilder.roles(roles);
```

How did I misunderstand your use case?

Reply:
sample:
``` java
public class SampleTests {
	@Mock
	MockHttpServletRequest request;
	
	private JwtRequestPostProcessor defaultTestAuthenticationPostProcessor() {
		return jwt().name("test_user").roles("a", "b");//more like claims here too
	}
	
	@Test
	public void jwtRolesAreAddedAndNotReplaced() {
		final JwtRequestPostProcessor rpp = defaultTestAuthenticationPostProcessor().roles("c");
		final JwtAuthenticationToken actual = (JwtAuthenticationToken) authentication(rpp.postProcessRequest(request));
		
		assertThat(actual.getAuthorities()).containsExactlyInAnyOrder(
				new SimpleGrantedAuthority("ROLE_a"),
				new SimpleGrantedAuthority("ROLE_b"),
				new SimpleGrantedAuthority("ROLE_c"));
	}

	static Authentication authentication(final MockHttpServletRequest req) {
		final SecurityContext securityContext = (SecurityContext) req.getAttribute(TestSecurityContextRepository.ATTR_NAME);
		return securityContext == null ? null : securityContext.getAuthentication();
	}
}
```
I'd expect to have "ROLE_C" only in that case, and it's the behavior I had implemented for all collections (roles, authorities, claims and scopes when it was still there). It should now behave as asserted in this test